### PR TITLE
Add basic http error handling.

### PR DESCRIPTION
**Pull request by ShimShtein**

### CODE REVIEW

File: app/jobs/send_notification_job.rb [hunk 1]
```diff
@@ -2,11 +2,32 @@
...
+  retry_on(Notifications::RecoverableError, wait: :exponentially_longer, attempts: 3) do |job, error|
+    endpoint = job.endpoint
+    disable_endpoint(endpoint)
+    Rails.logger.warn("Disabled #{endpoint} after too many retries for #{error.inspect}")
+  end
+
   def perform(endpoint, timestamp, category, message)
     Rails.logger.debug("Received a job to send: #{endpoint}, #{timestamp}, #{category}, #{message}")
 
     endpoint.send_message(timestamp: timestamp, category: category, message: message)
```
Comment (adamruzicka) [inline app/jobs/send_notification_job.rb]: If a delivery of message A gets queued, endpoint gets disabled then A gets processed, we would still try to deliver A even though the endpoint is disabled. Maybe we should check if the endpoint is active before hitting this?

File: app/jobs/send_notification_job.rb [hunk 2]
```diff
@@ -2,11 +2,32 @@
...
 # Sends a simple one-step notification.
 class SendNotificationJob < ApplicationJob
+  def self.disable_endpoint(endpoint)
+    endpoint.active = false
+    endpoint.save!
+  end
+
+  discard_on(Notifications::FatalError) do |job, error|
+    endpoint = job.endpoint
+    disable_endpoint(endpoint)
```
Comment (adamruzicka) [inline app/jobs/send_notification_job.rb]: Food for thought: if we disable an endpoint, none of scheduled deliveries after the current one matters anymore. We could leave things as they are and pop them one by one, or maybe we could try to wipe the entire queue.

Comment (adamruzicka) [reply app/jobs/send_notification_job.rb]: The previous comment was under the impression we use a queue-per-endpoint approach, which was in our early designs. Currently we use one common queue for all the endpoints and for that reason we can't just blow it away. Marking as resolved

File: .rubocop.yml [hunk 3]
```diff
@@ -23,3 +23,6 @@ Naming/FileName:
 
 Style/HashSyntax:
   Enabled: false
+
+Style/FormatStringToken:
```
[the remark this answers is not in the document]
Comment (ShimShtein) [reply .rubocop.yml]: Sorry, old version of exception handling. I'll remove it.

File: app/jobs/send_notification_job.rb [hunk 4]
```diff
@@ -2,11 +2,32 @@
...
+
   def perform(endpoint, timestamp, category, message)
     Rails.logger.debug("Received a job to send: #{endpoint}, #{timestamp}, #{category}, #{message}")
 
     endpoint.send_message(timestamp: timestamp, category: category, message: message)
 
     Rails.logger.debug("Job successfully sent: #{endpoint}, #{timestamp}, #{category}, #{message}")
   end
+
+  def endpoint
```
[the remark this answers is not in the document]
Comment (ShimShtein) [reply app/jobs/send_notification_job.rb]: This is the way I extract `endpoint` from `job` instance in [error handling](https://github.com/RedHatInsights/notifications-backend/pull/19/files/717e8225ed072898016a6012422f3702015cd765#diff-8b518916a449134183e4ac38fcb2341bR11)

File: app/jobs/send_notification_job.rb (hunk 1 above)
Comment (ShimShtein) [reply app/jobs/send_notification_job.rb]: I will add a condition to discard messages if the endpoint is not active

File: app/jobs/send_notification_job.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (adamruzicka) [reply app/jobs/send_notification_job.rb]: Oh, so the `job` passed into the error handling block is instance of the `SendNotificationJob` class. Then everything is fine

**Comment by adamruzicka** [comment]
Thanks @ShimShtein !

**Pull request merged**